core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java

```diff
@@ -275,7 +275,7 @@ public boolean allocatedPostIndexCreate(IndexMetaData indexMetaData) {
     }
 
     public boolean isPeerRecovery() {
-        return !primary() || relocatingNodeId() != null;
+        return state == ShardRoutingState.INITIALIZING && (primary() == false || relocatingNodeId != null);
```

Comment:
I prefer it this way as I see the `state == ShardRoutingState.INITIALIZING` condition as essential to both parts of the condition (it should also hold for non-primary shards). Making it 
`(state == ShardRoutingState.INITIALIZING && primary() == false) || isRelocationTarget()` does not read better...